src/stm32f1/quadrature_encoder.hpp

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <libhal-arm-mcu/stm32f1/timer.hpp>
+#include <libhal/pointers.hpp>
+#include <libhal/rotation_sensor.hpp>
+
+namespace hal::stm32f1 {
+/**
+ * @brief This class implements the `hal::rotation_sensor` interface
+ *
+ * It gets an input from a quadrature encoder motor and measures the amount turn
+ * using 2 channels as input.
+ *
+ * Each Quadrature Encoder must use channel 1 and 2 from the same timer.
+ */
+class quadrature_encoder : public hal::rotation_sensor
+{
+protected:
+  struct Key;
+
+public:
+  friend class hal::stm32f1::advanced_timer_manager;
+  friend class hal::stm32f1::general_purpose_timer_manager;
+  // this makes the constructor public, but due to the private Key member, only
+  // the timer_manager classes are allowed in call make_strong_ptr
```

Comment:
There isn't a reason to protect this constructor while its within the src directory. Headers here are not accessible by library uses, so you can eliminate the unnecessary key and privacy.